File: java/src/main/java/ru/fess38/finance/AppConfiguration.java
package ru.fess38.finance;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.TypeAdapterFactory;
import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;
import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;
import org.hibernate.SessionFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.io.ClassPathResource;
import org.springframework.http.converter.json.GsonHttpMessageConverter;
import org.springframework.orm.hibernate5.LocalSessionFactoryBean;
import org.springframework.scheduling.annotation.Scheduled;
import ru.fess38.finance.account.AccountBalanceCalculator;
import ru.fess38.finance.util.DatabaseEventListener;
import ru.fess38.finance.util.DefaultEntitiesCreator;
import ru.fess38.finance.util.DiskUtil;
import ru.fess38.finance.util.LocalDateConverter;

import java.time.LocalDate;
import java.util.ServiceLoader;
import javax.annotation.PostConstruct;
import javax.sql.DataSource;


@Configuration
public class AppConfiguration {
  private final Config config = ConfigFactory.load();
  private final DiskUtil diskUtil = new DiskUtil(config.getConfig("disk"));
  private DatabaseEventListener databaseEventListener;
  private DefaultEntitiesCreator defaultEntitiesCreator;
  private AccountBalanceCalculator balanceCalculator;

  @PostConstruct
  public void createDefaultEntities() {
    defaultEntitiesCreator.create();
  }

  @Scheduled(fixedRate = 600000, initialDelay = 60000)
  public void upload() throws Exception {
    diskUtil.upload(databaseEventListener.isChange());
    databaseEventListener.setChangeFalse();
  }

  @Scheduled(fixedRate = 10000)
  public void updateBalance() {
    if (databaseEventListener.isCalculateBalance()) {
      balanceCalculator.run();
    }
    databaseEventListener.setCalculateBalanceFalse();
  }

  @Bean
  public DatabaseEventListener databaseEventListener() {
    return new DatabaseEventListener();
  }

  @Bean
  public DataSource dataSource() throws Exception {
    diskUtil.download();
    HikariConfig hikariConfig = new HikariConfig();
    hikariConfig.setDataSourceClassName(config.getString("hikari.classname"));
    hikariConfig.addDataSourceProperty("url", config.getString("hikari.url"));
    return new HikariDataSource(hikariConfig);
  }

  @Bean
  @Autowired
  public SessionFactory sessionFactory(DataSource dataSource) throws Exception {
    LocalSessionFactoryBean factoryBean = new LocalSessionFactoryBean();
    factoryBean.setDataSource(dataSource);
    factoryBean.setConfigLocation(new ClassPathResource("ru/fess38/finance/hibernate.cfg.xml"));
    factoryBean.afterPropertiesSet();
    return factoryBean.getObject();
  }

  @Bean
  public GsonHttpMessageConverter gsonHttpMessageConverter() {
    GsonBuilder gsonBuilder = new GsonBuilder();
    for (TypeAdapterFactory factory : ServiceLoader.load(TypeAdapterFactory.class)) {
      gsonBuilder.registerTypeAdapterFactory(factory);
    }
    gsonBuilder.registerTypeAdapter(LocalDate.class, new LocalDateConverter());
    Gson gson = gsonBuilder.create();
    GsonHttpMessageConverter gsonHttpMessageConverter = new GsonHttpMessageConverter();
    gsonHttpMessageConverter.setGson(gson);
    return gsonHttpMessageConverter;
  }

  @Autowired
  public void setDatabaseEventListener(DatabaseEventListener databaseEventListener) {
    this.databaseEventListener = databaseEventListener;
  }

  @Autowired
  public void setDefaultEntitiesCreator(DefaultEntitiesCreator defaultEntitiesCreator) {
    this.defaultEntitiesCreator = defaultEntitiesCreator;
  }

  @Autowired
  public void setBalanceCalculator(AccountBalanceCalculator balanceCalculator) {
    this.balanceCalculator = balanceCalculator;
  }
}

File: kotlin/src/main/kotlin/ru/fess38/finance/AppConfiguration.kt
package ru.fess38.finance

import com.google.api.client.googleapis.auth.oauth2.GoogleIdTokenVerifier
import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport
import com.google.api.client.json.jackson2.JacksonFactory
import com.google.protobuf.Message
import com.typesafe.config.Config
import com.typesafe.config.ConfigFactory
import org.apache.commons.dbcp2.BasicDataSource
import org.hibernate.SessionFactory
import org.springframework.boot.autoconfigure.web.HttpMessageConverters
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.converter.protobuf.ProtobufHttpMessageConverter
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.orm.hibernate5.HibernateTransactionManager
import org.springframework.orm.hibernate5.LocalSessionFactoryBean
import org.springframework.transaction.PlatformTransactionManager
import ru.fess38.finance.core.MessageService
import ru.fess38.finance.repository.HibernateEntity
import ru.fess38.finance.security.User
import ru.fess38.finance.validation.CompositeValidator
import ru.fess38.finance.validation.MessageValidator
import java.util.Properties
import javax.sql.DataSource

@Configuration
class AppConfiguration {
  @Bean
  fun config() = ConfigFactory.load().getConfig(System.getenv("env") ?: "dev")!!

  @Bean
  fun dataSource(config: Config): DataSource {
    val dataSource = BasicDataSource()
    dataSource.driverClassName = config.getString("postgres.classname")
    val host = config.getString("postgres.host")
    val port = config.getString("postgres.port")
    val database = config.getString("postgres.databasename")
    dataSource.url = "jdbc:postgresql://$host:$port/$database"
    dataSource.username = config.getString("postgres.user")
    dataSource.password = config.getString("postgres.password")
    dataSource.validationQuery = "SELECT version();"
    return dataSource
  }

  @Bean
  fun sessionFactory(dataSource: DataSource, config: Config): SessionFactory {
    val factoryBean = LocalSessionFactoryBean()
    factoryBean.setDataSource(dataSource)
    factoryBean.hibernateProperties = config.getConfig("hibernate").toProperties()
    factoryBean.setAnnotatedClasses(HibernateEntity::class.java, User::class.java)
    factoryBean.afterPropertiesSet()
    return factoryBean.`object`
  }

  @Bean
  fun jdbcTemplate(dataSource: DataSource) = JdbcTemplate(dataSource)

  @Bean
  fun transactionManager(sessionFactory: SessionFactory): PlatformTransactionManager {
    return HibernateTransactionManager(sessionFactory)
  }

  @Bean
  fun protobufHttpMessageConverter(): ProtobufHttpMessageConverter {
    return ProtobufHttpMessageConverter()
  }

  @Bean
  fun httpMessageConverters(converter: ProtobufHttpMessageConverter): HttpMessageConverters {
    return HttpMessageConverters(converter)
  }

  @Bean
  fun googleIdTokenVerifier(config: Config) = GoogleIdTokenVerifier
      .Builder(GoogleNetHttpTransport.newTrustedTransport(), JacksonFactory.getDefaultInstance())
      .setAudience(listOf(config.getString("security.google.clientId")))
      .build()!!

  @Bean
  fun messageValidator(messageService: MessageService): MessageValidator<Message> {
    return CompositeValidator(messageService)
  }
}

fun Config.toProperties(): Properties {
  val properties = Properties()
  this.entrySet().forEach {properties[it.key] = it.value.unwrapped()}
  return properties
}
